pytest-embedded-idf/pytest_embedded_idf/idf_unity_tester.py

```diff
@@ -86,19 +100,20 @@ def run_normal_case(self, case: UnittestMenuCase, reset: bool = False, timeout:
             case: the specific case that parsed in test menu
             reset: whether do a hardware reset before running the case
             timeout: timeout in second
+            retry_count (int): number of retries for a single case when it is failed to start
         """
         if case.type == 'normal':
             if reset:
                 self.dut.serial.hard_reset()
             self.dut.expect(self.ready_pattern_list, timeout=timeout)
             # Retry if write not success
-            for retry in range(self.retry_times):
+            for retry in range(retry_count):
                 self.dut.write(str(case.index))
                 try:
-                    self.dut.expect('Running {}...'.format(case.name), timeout=1)
+                    self.dut.expect(f'Running {case.name}...', timeout=1)
                     break
                 except TIMEOUT as e:
-                    if retry >= self.retry_times - 1:
+                    if retry >= retry_count - 1:
                         raise e
             self.dut.expect_unity_test_output(timeout=timeout)
```

Comment:
should also retry if this test case failed. here's the logic is a bit more complicated.

- if the result is passed:
    return success
- if the result is failed (or no result since core dumped or whatever) but retried:
    not decided yet
- if the result is failed (or no result since core dumped or whatever) and the retry_count has exceeded:
   return failed

Reply:
this try...except... block should wrap the whole block, like

```python
try:
    self.dut.write(str(case.index))
    self.dut.expect(f'Running {case.name}...', timeout=1)
    self.dut.expect_unity_test_output(timeout=timeout)
except ...:
    ...
```

besides, the timeout check should also calculate the whole process., use some techniques like https://docs.python.org/3/library/threading.html#threading.Timer

Reply:
same logic to other `run_...` functions

Reply:
I suppose this logic can be applied to the `run_normal_case` only. For `run_multistage` and `run_multidev` it requires to pass the `sub_case_index`, so the `dut.write(str(sub_case_index))` can't be placed under `try` block

Reply:
what do you mean by "can't be placed under the try block" since those two functions "requires to pass the sub_case_index"?

Reply:
> 



> this try...except... block should wrap the whole block, like
> 
> ```python
> try:
>     self.dut.write(str(case.index))
>     self.dut.expect(f'Running {case.name}...', timeout=1)
>     self.dut.expect_unity_test_output(timeout=timeout)
> except ...:
>     ...
> ```
> 
> besides, the timeout check should also calculate the spent time of the code block above. You may use https://docs.python.org/3/library/threading.html#threading.Timer or anything similar.

I would suggest to remain the legacy logic. Reasons given in https://github.com/espressif/pytest-embedded/pull/153#discussion_r1009061388

Reply:
I think we're talking about two things here... Let me clarify it a bit:

- for the idea to support keywords defined with `TEST_CASE` (described here: https://github.com/espressif/pytest-embedded/pull/153#discussion_r1009090953), please add the retry mechanism to the whole block
- for the idea to run retry while starting the test case, I agree with @L-KAYA 's idea. We could improve it later in a separate PR.